### DISCUSSION TITLE
Feature request: recognize server-only folders

### DISCUSSION THREAD

**Discussion by jonschlinkert** [remix-run/remix#4093] [Proposals]
I looked through the docs, issues and discussions, and didn't see anything that supported this. 

To simplify adding and organizing server-only code, it would be awesome if remix had built-in support for a `server` folder, where all files inside that folder do not require the `.server.*` suffix. 

Is this possible currently? If not, perhaps we can add support for server-only glob patterns in `remix.config.js`? 

edit: the closest thing I saw was [ignoredroutefiles](https://remix.run/docs/en/v1/api/conventions#ignoredroutefiles). Is that what I'm looking for? My assumption was that ignored routes would still be bundled with client code. Apologies if I misunderstood. 

I also want to mention that I think the convention of using `*.server.ts` is great. I don't think having support for a `server` folder will confuse people or create a reason not to use `*.server.ts` files in general.

**Comment by pawelblaszczyk5** [comment]
Currently it's not possible within config, you can use this [patch](https://github.com/kiliman/remix-server-folders) by @kiliman though

**Comment by jonschlinkert** [comment]
This also makes me curious about how `index.*` files are currently treated by remix. For example, given the following: 

```
app/
├── models/
│   ├── one.server.ts
│   ├── two.server.ts
│   └── index.ts
└── root.tsx
```
Is `index.ts` still considered client-side code, even though all of the files in `models/` have the `.server.ts` suffix? I'm guessing that remix probably tree-shakes those server-side paths out of the index file first, then `index.ts` ends up being empty anyway.

**Reply by pawelblaszczyk5** [reply]
Afaik you're guessing correctly, I used barrel files with *.server files and it worked :)

**Reply by runofthemill** [reply]
I'm curious if this is still true for you - I'm trying it out, and I have a utility file w/ some references to `process.env`. When the file is named `index.ts`, I get an error in the browser console even though the imported code is only used in loaders:

```
index.ts:23 Uncaught ReferenceError: process is not defined
    at index.ts:23:7
```

Renaming to `index.server.ts` fixes the issue, as expected, **buuuut** if I change it to a barrel import, e.g. move the code from `index.ts` to `utils.server.ts`, and export from `index.ts`:

```ts
// index.ts
export * from "./utils.server"
```

I get the uncaught reference error again:

```
utils.server.ts:23 Uncaught ReferenceError: process is not defined
    at utils.server.ts:23:7
```

(Also looks like the barrel import file was removed, but that seems to make sense)

**Comment by atul-jalan** [comment]
This would be hugely useful for me too! As my projects scale, its useful to co-locate lots of server files in one location (e.g. for third party integrations), and it'd be nice to not have to add a `.server` to every filename.

**Comment by loiclefloch** [comment]
This is something we have think about for some time too.

Our codebase is separated in different layers (dto / services / dao). 

If this feature request is accepted, I would gladly help to implement it, if needed.

Thank you !

**Comment by 921kiyo** [comment]
I am wondering if server-only folder is already supported (if so, from which version?). I just looked at the release note, but is it done with the combination of Vite (I am a bit new to Vite/Remix ecosystem, so apologies in advance if I am asking something obvious)?

**Reply by kiliman** [reply]
Yes, this is now part of Remix v2.7, but _only_ using Vite. The current Remix compiler still requires filenames to include the `.server` suffix.
